### PR TITLE
DM-24389: Implement amp-to-amp offset correction task

### PR DESCRIPTION
**Pull request by leeskelvin**
HSC imaging is known to exhibit subtle 1-2 count amp-to-amp offsets.
This commit implements an empirical amp-to-amp offset correction
technique within ISR processing, removing these offset artifacts.

### CODE REVIEW

File: python/lsst/obs/subaru/ampOffset.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+                edgeB = np.nanmedian(stripB, axis=1)
+            edgeDiff = edgeB - edgeA
+            # compute rolling averages
+            edgeDiffSum = np.convolve(np.nan_to_num(edgeDiff), np.ones(self.config.ampEdgeWindow), 'same')
+            edgeDiffNum = np.convolve(~np.isnan(edgeDiff), np.ones(self.config.ampEdgeWindow), 'same')
+            edgeDiffAvg = edgeDiffSum / np.clip(edgeDiffNum, 1, None)
+            edgeDiffAvg[np.isnan(edgeDiff)] = np.nan
+            # take clipped mean of rolling average data as amp-edge step value
+            ampStep = afwMath.makeStatistics(edgeDiffAvg, afwMath.MEANCLIP, sctrl).getValue()
+            minFracBad = (1 - (np.sum(np.isnan(edgeDiffAvg))/len(edgeDiffAvg))) < self.config.ampEdgeMinFrac
```
Comment (erykoff) [inline python/lsst/obs/subaru/ampOffset.py:124]: Add a couple of comments on what's being checked here.  (a) that the fraction of pixels is large enough (?) and (b) that the step size is less than the configured max.

File: python/lsst/obs/subaru/ampOffset.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+class SubaruAmpOffsetConfig(AmpOffsetConfig):
+    """Configuration parameters for SubaruAmpOffsetTask.
+    """
+
+    def setDefaults(self):
+        self.ampEdgeInset = 5
+        self.ampEdgeWidth = 64
+        self.ampEdgeMinFrac = 0.5
+        self.ampEdgeMaxStep = 5.0
+        self.ampEdgeWindow = 512
```
[the remark this answers is not in the document]
Comment (leeskelvin) [reply python/lsst/obs/subaru/ampOffset.py]: Thank you, these have been removed as duplicates.

File: python/lsst/obs/subaru/ampOffset.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        table = afwTable.SourceTable.make(schema)
+        # detection sigma, used for smoothing and to grow detections, is
+        # normally measured from the PSF of the exposure. As the PSF has not
+        # yet been measured at this stage of processing, sigma is instead set
+        # to an approximate value here which should be sufficient.
+        _ = self.detection.run(table=table, exposure=exp, sigma=2)
+
+        # safety check: do any pixels remain for amp offset estimation?
+        if (exp.mask.array & bitMask).all():
+            raise pipeBase.TaskError("All pixels masked: cannot calculate any amp offset corrections.")
```
[the remark this answers is not in the document]
Comment (leeskelvin) [reply python/lsst/obs/subaru/ampOffset.py]: This is now a warning, i.e., `warnings.warn("All pixels masked: cannot calculate any amp offset corrections.")`

File: python/lsst/obs/subaru/ampOffset.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+        # set up amp offset inputs
+        im = exp.image
+        im.array[(exp.mask.array & bitMask) > 0] = np.nan
+        amps = exp.getDetector().getAmplifiers()
+        ampEdgeOuter = self.config.ampEdgeInset + self.config.ampEdgeWidth
+        sctrl = afwMath.StatisticsControl()
+
+        # loop over each amp edge boundary to extract amp-edge step values
```
[the remark this answers is not in the document]
Comment (leeskelvin) [reply python/lsst/obs/subaru/ampOffset.py]: This now reads `# loop over each amp edge boundary to extract amp offset values`

File: python/lsst/obs/subaru/ampOffset.py (hunk 1 above)
Comment (leeskelvin) [reply python/lsst/obs/subaru/ampOffset.py]: The following comments have been added here:
```
# perform a couple of do-no-harm safety checks:
# a) the fraction of unmasked pixel rows is > ampEdgeMinFrac,
# b) the absolute offset ADU value is < ampEdgeMaxOffset
```

File: config/isr.py [hunk 5]
```diff
@@ -112,6 +113,9 @@
 
 config.doNanInterpAfterFlat = False
 
+config.doAmpOffset = True
```
[the remark this answers is not in the document]
Comment (leeskelvin) [reply config/isr.py]: Updated.

**Review by erykoff — APPROVED** [review]
This looks great!  I do have some smallish comments, and I hope we get to turn this on at some point soon!

File: python/lsst/obs/subaru/ampOffset.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    def setDefaults(self):
+        # the binSize should be as large as possible, preventing background
+        # subtraction from inadvertently removing the amp offset signature.
+        # Here it's set to the width of an HSC amp, which seems reasonable.
+        self.background.binSize = 512
+        self.background.algorithm = "AKIMA_SPLINE"
+        self.background.useApprox = False
+        self.background.ignoredPixelMask = ["BAD", "SAT", "INTRP", "CR", "EDGE", "DETECTED",
+                                            "DETECTED_NEGATIVE", "SUSPECT", "NO_DATA"]
+        self.detection.reEstimateBackground = False
```
[the remark this answers is not in the document]
Comment (leeskelvin) [reply python/lsst/obs/subaru/ampOffset.py:43]: When I try to change it to `reestimateBackground`, I get a `*** AttributeError: lsst.meas.algorithms.detection.SourceDetectionConfig has no attribute reestimateBackground` - from what I can tell, `reEstimateBackground` seems to be correct - can you confirm?

[the remark this answers is not in the document]
Comment (erykoff) [reply python/lsst/obs/subaru/ampOffset.py:43]: Oh! Sorry, this isn't your config, this is somebody else with a bad name?  Well, never mind then.

File: python/lsst/obs/subaru/ampOffset.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+            # solve for pedestal values and update original exposure in-place
+            A = np.array([[-1.0, 1.0, 0.0, 0.0],
+                          [1.0, -2.0, 1.0, 0.0],
+                          [0.0, 1.0, -2.0, 1.0],
+                          [0.0, 0.0, 1.0, -1.0]])
+            B = np.array([ampOffsets[0],
+                          ampOffsets[1] - ampOffsets[0],
+                          ampOffsets[2] - ampOffsets[1],
+                          -ampOffsets[2]])
+            pedestals = np.nan_to_num(np.linalg.lstsq(A, B, rcond=None)[0])
```
[the remark this answers is not in the document]
Comment (leeskelvin) [reply python/lsst/obs/subaru/ampOffset.py:145]: Yes, you're correct in that if the least squares minimization fails for whatever reason, returning NaN's, the `nan_to_num` wrapper will convert these to zeroes, i.e., nothing will be added or subtracted from any of the four pedestals. A comment has been added in-line to reflect this.

File: python/lsst/obs/subaru/ampOffset.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+class SubaruAmpOffsetTask(AmpOffsetTask):
+    """Calculate and apply amp offset corrections to an exposure.
+    """
+    ConfigClass = SubaruAmpOffsetConfig
+    _DefaultName = "subaruIsrAmpOffset"
+
+    def __init__(self, *args, **kwargs):
+        AmpOffsetTask.__init__(self, *args, **kwargs)
+        # always load background subtask, even if doBackground=False;
```
[the remark this answers is not in the document]
Comment (leeskelvin) [reply python/lsst/obs/subaru/ampOffset.py]: Moved to the parent task, as suggested.

File: python/lsst/obs/subaru/ampOffset.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+            bgImage = bg.getImageF(self.background.config.algorithm, self.background.config.undersampleStyle)
+            maskedImage -= bgImage
+
+        # detect sources and update cloned exposure mask planes in-place
+        if self.config.doDetection:
+            schema = afwTable.SourceTable.makeMinimalSchema()
+            table = afwTable.SourceTable.make(schema)
+            # detection sigma, used for smoothing and to grow detections, is
+            # normally measured from the PSF of the exposure. As the PSF hasn't
+            # been measured at this stage of processing, sigma is  instead
```
[the remark this answers is not in the document]
Comment (leeskelvin) [reply python/lsst/obs/subaru/ampOffset.py]: Removed, thanks!

**Pull request merged**